pypicloud/storage/gcs.py

```diff
@@ -172,8 +172,22 @@ def list(self, factory=Package):
         for blob in blobs:
             pkg = self.package_from_object(blob, factory)
             if pkg is not None:
+                # If we have a separate upload prefix, flag THIS package as being a fallback. Otherwise
+                # we don't know enough to differentiate.
+                pkg.origin = "fallback" if self.upload_prefix else None
                 yield pkg
 
+        # If we have an upload_prefix, now go back and process anything that matches.
+        if self.upload_prefix:
+            blobs = self.bucket.list_blobs(prefix=self.upload_prefix)
+            for blob in blobs:
+                pkg = self.package_from_object(blob, factory)
+                if pkg is not None:
+                    # If we have a separate upload prefix, flag THIS package as being a fallback. Otherwise
```

Comment:
should be "flag THIS package as being an _upload_" rather than "fallback"

Reply:
Actually it's correct. We use the presence of an upload_prefix as the discriminator to determine whether the differentiation feature is even active. If we are in the default mode of non-discrimination, we don't set pkg.origin at all (even though the result would simply be that all packages would be set to "fallback").